Bluetooth: controller: Fix ticker catch up on ISR latency

cvinayak opened the pull request:
Fix ticker to avoid catch up of periodic timeouts in case of
large ISR latencies like in case of flash erase scenarios.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
Bluetooth: controller: Fix ticker catch up on ISR latency

Fix ticker to avoid catch up of periodic timeouts in case of
large ISR latencies like in case of flash erase scenarios.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
Bluetooth: controller: nRF: Revert use of ticker compat mode as default

Ticker is now fixed to avoid catch up of periodic timeout under
large ISR latencies.

Revert commit a749e28d986c ("Bluetooth: controller: split:
nRF: Use ticker compat mode as default").

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

mtpr-ot commented on subsys/bluetooth/controller/ticker/ticker.c:
```diff
@@ -1433,11 +1439,46 @@ static inline void ticker_job_worker_bh(struct ticker_instance *instance,
+					ticks_to_expire +=
+						ticker_remainder_inc(ticker);
+					lazy++;
+				}
```
I assume this is a safeguard for when lagging behind, and catching up?

cvinayak replied:
Good point, will fix this to consider the "must_expire" ticker instances.

mtpr-ot reviewed (changes requested):
As I see it, this change will make ticker_job take over the lazy/latency handling, which is fine and more efficient, but then the lazy handling in ticker_worker should be removed (a few lines).

cvinayak commented:
> As I see it, this change will make ticker_job take over the lazy/latency handling, which is fine and more efficient, but then the lazy handling in ticker_worker should be removed (a few lines).

Thanks for pointing out, will address this.

cvinayak committed:
Bluetooth: controller: Remove lazy handling in ticker_worker

Removed lazy handling from ticker_worker as it is now taken
care in ticker_job.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
Bluetooth: controller: Consider must_expire while avoiding catchup

If must_expire is set for a ticker instance, then ticker
expiry shall still perform catchup.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
Bluetooth: controller: Fix ticker state on skipped

Reset ticker state in ticker_job for ticker instances that
have been skipped in the ticker_worker.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
Bluetooth: controller: Fix ticks_slot_previous calculation

Fix ticks_slot_previous calculation in the ticker_job when
tickers are skipped.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak commented:
@mtpr-ot I have reworked, some more fixes, and they are in additional separate commits in this PR. Please review.

mtpr-ot commented:
> @mtpr-ot I have reworked, some more fixes, and they are in additional separate commits in this PR. Please review.

Working on it - I think I need to test it, I'm a bit unsure about impact of the latest change. Feedback later today

mtpr-ot commented:
Testing with this PR shows problems honoring critical priority events (priority = -128), getting shallow "must_expire" timeouts where they should go in the air. Will continue to investigate.

mtpr-ot commented on subsys/bluetooth/controller/ticker/ticker.c:
```diff
@@ -1388,32 +1387,25 @@ static inline void ticker_job_worker_bh(struct ticker_instance *instance,
-		if (ticker->lazy_current != 0U &&
-		    !TICKER_RESCHEDULE_PENDING(ticker)) {
+		ticks_latency -= ticks_to_expire;
+#endif /* !CONFIG_BT_TICKER_COMPATIBILITY_MODE */
```
I think we need to keep the logic that if lazy_current != 0 and no pending reschedules, then clear the slot_previous info and nothing else. This is needed because must_expire still needs special handling for must_expired nodes that don't go in the air, and catch a lazy count. By checking for this in addition, I think we get the desired functionality for both "worlds". At least my tests pass with this:
```
#if !defined(CONFIG_BT_TICKER_COMPATIBILITY_MODE)
		ticks_latency -= ticks_to_expire;

		if (ticker->must_expire && ticker->lazy_current != 0U &&
		    !TICKER_RESCHEDULE_PENDING(ticker)) {
			instance->ticker_id_slot_previous = TICKER_NULL;
			instance->ticks_slot_previous = 0U;
		} else
#endif /* !CONFIG_BT_TICKER_COMPATIBILITY_MODE */
		{
```

cvinayak replied:
A skipped ticker instance shall not clear the `ticks_slot_previous`, wouldn't that also apply to a "must expire" tickers? We should have a audio call at the next opportunity to review this.

mtpr-ot replied:
Yes, you're right. I had a sneaking suspicion after writing that, and went back to the code this morning. I will take a second look, and let's talk about this Monday morning, if you're available.

mtpr-ot replied:
I think I figured it out. The problem is in the updating of slot_previous to the expired ticker id, regardsless of must_expire:
```
	/* If a reschedule is set pending, we will need to keep
	 * the slot_previous information
	 */
	if ((ticker->ticks_slot != 0U) &&
	    !(ticker->must_expire && ticker->lazy_current != 0U) && // <----- added
	    (((ticker->req - ticker->ack) & 0xff) == 2U) &&
	    !TICKER_RESCHEDULE_PENDING(ticker)) {
		instance->ticker_id_slot_previous =
			id_expired;
		instance->ticks_slot_previous =
			ticker->ticks_slot;
	}
```
This will overwrite the info from the blocking node, and use the ticks_slot from the must_expire (shallow) expiry instead, which is incorrect.

cvinayak replied:
I have tested and added your check, PR updated.

mtpr-ot replied:
That's great. Doing final re-test here. However, unfortunately you need to apply conditional compile with CONFIG_BT_TICKER_COMPATIBILITY_MODE around that added line :-/

cvinayak replied:
my fault, I was aware, but forgot to add the IS_ENABLE(...). Will handle and push updates soon.

mtpr-ot reviewed (approved):
LGTM

thoh-ot commented:
Please create a backport pr for this.
